Add configurable control/tunnel addresses for the server

confor opened the pull request:
I've added the option to bind to the given IP address for both the control server and the tunnels, which may be different. Previously both were hardcoded to listen on `0.0.0.0`.

A new `--control-addr` flag allows users to specify the IP address where the control server will listen on. Likewise for `--tunnels-addr` allowing users to define the IP address on which tunnels will listen.

Its now possible to isolate the control server from the tunnel's network. One could perhaps have the control server on a `tailscale0` interface while the tunnels listen on `eth0` on a public IP.

I have already tested bore with this change to redirect a nginx instance (server @ `10.0.0.95:80`) via tailscale (control @ `100.64.0.3`) into another interface (tunnel @ `192.168.30.100`) and it works fine on my machines.

ekzhang commented on src/main.rs:
```diff
@@ -47,6 +47,14 @@ enum Command {
         #[clap(short, long, env = "BORE_SECRET", hide_env_values = true)]
         secret: Option<String>,
+
+        /// IP address for the control server. Bore clients must reach this address.
+        #[clap(long, default_value = "0.0.0.0")]
+        control_addr: String,
+
+        /// IP address where tunnels will listen on.
+        #[clap(long, default_value = "0.0.0.0")]
+        tunnels_addr: String,
```
Could you make their type `IpAddr` for validation, and clarify in the name or comment that it's the "Bind address"? This might make it clearer the behavior of something like `0.0.0.0` or `::`.

Would also be nice to have `tunnels_addr` be an optional argument type so that if it's not specified, it's the same as `control_addr`. Since I imagine setting both to the same value would be the most common use case, e.g., adjusting it to IPv6.

Something like

```bash
bore server --bind-addr ::

bore server --bind-addr :: --bind-tunnels b0ef:cfe5:5952:6f92:8687:49c1:15f4:fae5
```

confor replied:
I didn't know that I could just use IpAddr directly as the clap type :o

ekzhang commented on src/server.rs:
```diff
@@ -23,25 +23,32 @@ pub struct Server {
-        let listener = TcpListener::bind(&addr).await?;
-        info!(?addr, "server listening");
+        let listener = TcpListener::bind((this.control_addr.as_ref(), CONTROL_PORT)).await?;
+        info!(?this.control_addr, "server listening");
```
```suggestion
        info!(addr = ?this.control_addr, "server listening");
```

ekzhang reviewed (commented):
Lgtm, nice change and very minimal! Very thoughtful to split up the two bind addresses, and this will support ipv6 as well.

Could you take a look at the suggestions?

confor commented:
I've added IpAddr with descriptive error messages and renamed the two things.

What's the proper way to default an option to another option? I was thinking something simple like:

```rust
if bind_tunnels.is_none() {
    Server::new(port_range, secret.as_deref(), bind_addr, bind_addr) ...
} else {
    Server::new(port_range, secret.as_deref(), bind_addr, bind_tunnels) ...
}
```

Not sure if clap has something like `default_value = self.bind_addr` for this

ekzhang commented:
> What's the proper way to default an option to another option?

`let bind_tunnels = bind_tunnels.unwrap_or(bind_addr);` maybe?

confor commented:
done! I also added the address to the printf:
```
$ ./bore server --bind-addr 127.0.0.64
2025-04-14T21:21:43.671724Z  INFO bore_cli::server: server listening addr=127.0.0.64 port=7835
2025-04-14T21:21:45.531123Z  INFO control{addr=127.0.0.1:45086}: bore_cli::server: incoming connection
2025-04-14T21:21:45.531185Z  INFO control{addr=127.0.0.1:45086}: bore_cli::server: new client host=127.0.0.64 port=61759

```

```
$ ./bore local 8080 --to 127.0.0.64
2025-04-14T21:21:45.531246Z  INFO bore_cli::client: connected to server remote_port=61759
2025-04-14T21:21:45.531255Z  INFO bore_cli::client: listening at 127.0.0.64:61759
```
